### ISSUE TITLE
ossec-modulesd SCA CIS check fails for directory entry (cis_debian10 rule 2526)

### ISSUE THREAD

**Issue by fraform-tnz** [wazuh/wazuh#9069]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.1.5-1 | ossec-modulesd | Agent | Packages | Debian 10 |

Hi all, I have a problem with evaluation of SCA CIS rule 2526 for cis_debian10.

![wazuh](https://user-images.githubusercontent.com/59601514/122738568-65c62d00-d282-11eb-8405-e358c5edf437.png)

I have created a file named **cis_sudoers** in /etc/sudoers.d folder as suggested in **Remediation** section of CIS benchmark guideline but the check fails.

Here is the log of /var/ossec/bin/ossec-modulesd -dd -f:

```
wm_sca.c:1448 at wm_sca_check_file_contents(): DEBUG: Result for (r:^\s*\t*Defaults\s*\t*use_pty)(/etc/sudoers) -> 0
wm_sca.c:1539 at wm_sca_check_file_list_for_contents(): DEBUG: Match not found in file '/etc/sudoers'. Continuing.
wm_sca.c:1543 at wm_sca_check_file_list_for_contents(): DEBUG: Result for (r:^\s*\t*Defaults\s*\t*use_pty)(/etc/sudoers) -> 0
wm_sca.c:1157 at wm_sca_do_scan(): DEBUG: Result for rule 'f:/etc/sudoers -> r:^\s*\t*Defaults\s*\t*use_pty': 0
wm_sca.c:1004 at wm_sca_do_scan(): DEBUG: Considering rule: 'd:/etc/sudoers.d -> r:\.* -> r:^\s*\t*Defaults\s*\t*use_pty'
wm_sca.c:1103 at wm_sca_do_scan(): DEBUG: Processing directory rule '/etc/sudoers.d -> r:\.* -> r:^\s*\t*Defaults\s*\t*use_pty'
wm_sca.c:812 at wm_sca_check_dir_list(): DEBUG: Exploring directories [/etc/sudoers.d]
wm_sca.c:815 at wm_sca_check_dir_list(): DEBUG: Checking directory '/etc/sudoers.d' => is_nfs=0, skip_nfs=1
wm_sca.c:1879 at wm_sca_check_dir(): DEBUG: Checking directory '/etc/sudoers.d' -> r:\.* -> r:^\s*\t*Defaults\s*\t*use_pty
wm_sca.c:779 at wm_sca_resolve_symlink(): DEBUG: Resolving real path of '/etc/sudoers.d'
wm_sca.c:799 at wm_sca_resolve_symlink(): DEBUG: Real path of '/etc/sudoers.d' is '/etc/sudoers.d'
wm_sca.c:1923 at wm_sca_check_dir(): DEBUG: Considering directory entry '/etc/sudoers.d/cis_sudoers'
wm_sca.c:1944 at wm_sca_check_dir(): DEBUG: Skipping directory entry '/etc/sudoers.d/cis_sudoers'
wm_sca.c:1960 at wm_sca_check_dir(): DEBUG: Check result for dir '/etc/sudoers.d': 0
wm_sca.c:1118 at wm_sca_do_scan(): DEBUG: Check directory rule result: 0
wm_sca.c:1157 at wm_sca_do_scan(): DEBUG: Result for rule 'd:/etc/sudoers.d -> r:\.* -> r:^\s*\t*Defaults\s*\t*use_pty': 0
wm_sca.c:1180 at wm_sca_do_scan(): DEBUG: Result for check id: 2526 'Ensure sudo commands use pty' -> 0
```

I also checked at source code of wm_sca.c file (https://github.com/wazuh/wazuh/blob/46f640b8684136047e8e6b0af3e0ca034b863f5f/src/wazuh_modules/wm_sca.c) at lines 2004-2013 and I think the problem could be at line 2006 but really I don't know how to test it (i checked only first part of "if" statement but no OS_Regex and OS_Match statements).

```
2004:        if (S_ISDIR(statbuf_local.st_mode)) {
2005:            result = wm_sca_check_dir(f_name, file, pattern, reason);
2006:        } else if (((file && strncasecmp(file, "r:", 2) == 0) && OS_Regex(file + 2, entry->d_name))
2007:                || OS_Match2(file, entry->d_name))
2008:        {
2009:            result = wm_sca_check_file_list(f_name, pattern, reason);
2010:        } else {
2011:            mdebug2("Skipping directory entry '%s'", f_name);
2012:            continue;
2013:        }
```

Moreover I tested the regex with /var/ossec/bin/ossec-regex and it seems to work

```
/var/ossec/bin/ossec-regex '^\s*\t*Defaults\s*\t*use_pty'
Defaults    use_pty
+OSRegex_Execute: Defaults    use_pty
+OS_Regex       : Defaults    use_pty
```

**Comment by chemamartinez** [comment]
Hi @fraform-tnz,

Thank you for your detailed report.

Regarding the debug output you show, it seems the SCA scan is skipping the file `cis_sudoers`:

```
wm_sca.c:1923 at wm_sca_check_dir(): DEBUG: Considering directory entry '/etc/sudoers.d/cis_sudoers'
wm_sca.c:1944 at wm_sca_check_dir(): DEBUG: Skipping directory entry '/etc/sudoers.d/cis_sudoers'
```

However, the same check is working fine for me, here you can see the debug output of the same check when created the file `cis_sudoers` in `/etc/sudoers.d` with the content `Defaults use_pty`:

```
2021/06/25 08:25:31 sca[5640] wm_sca.c:1990 at wm_sca_check_dir(): DEBUG: Considering directory entry '/etc/sudoers.d/cis_sudoers'
2021/06/25 08:25:31 sca[5640] wm_sca.c:1568 at wm_sca_check_file_list_for_contents(): DEBUG: Checking file list '/etc/sudoers.d/cis_sudoers' with 'r:^\s*\t*Defaults\s*\t*use_pty'
2021/06/25 08:25:31 sca[5640] wm_sca.c:794 at wm_sca_resolve_symlink(): DEBUG: Resolving real path of '/etc/sudoers.d/cis_sudoers'
2021/06/25 08:25:31 sca[5640] wm_sca.c:814 at wm_sca_resolve_symlink(): DEBUG: Real path of '/etc/sudoers.d/cis_sudoers' is '/etc/sudoers.d/cis_sudoers'
2021/06/25 08:25:31 sca[5640] wm_sca.c:1456 at wm_sca_check_file_existence(): DEBUG: FILE_EXISTS(/etc/sudoers.d/cis_sudoers) -> RETURN_FOUND
2021/06/25 08:25:31 sca[5640] wm_sca.c:1471 at wm_sca_check_file_contents(): DEBUG: Checking contents of file '/etc/sudoers.d/cis_sudoers' against pattern 'r:^\s*\t*Defaults\s*\t*use_pty'
2021/06/25 08:25:31 sca[5640] wm_sca.c:794 at wm_sca_resolve_symlink(): DEBUG: Resolving real path of '/etc/sudoers.d/cis_sudoers'
2021/06/25 08:25:31 sca[5640] wm_sca.c:814 at wm_sca_resolve_symlink(): DEBUG: Real path of '/etc/sudoers.d/cis_sudoers' is '/etc/sudoers.d/cis_sudoers'
2021/06/25 08:25:31 sca[5640] wm_sca.c:1900 at wm_sca_pattern_matches(): DEBUG: Testing minterm (r:^\s*\t*Defaults\s*\t*use_pty)(Defaults use_pty) -> 1
2021/06/25 08:25:31 sca[5640] wm_sca.c:1903 at wm_sca_pattern_matches(): DEBUG: Pattern test result: (r:^\s*\t*Defaults\s*\t*use_pty)(Defaults use_pty) -> 1
2021/06/25 08:25:31 sca[5640] wm_sca.c:1506 at wm_sca_check_file_contents(): DEBUG: (r:^\s*\t*Defaults\s*\t*use_pty)(Defaults use_pty) -> 1
2021/06/25 08:25:31 sca[5640] wm_sca.c:1509 at wm_sca_check_file_contents(): DEBUG: Match found. Skipping the rest.
2021/06/25 08:25:31 sca[5640] wm_sca.c:1515 at wm_sca_check_file_contents(): DEBUG: Result for (r:^\s*\t*Defaults\s*\t*use_pty)(/etc/sudoers.d/cis_sudoers) -> 1
```

It matches and the check result is passed.

Could you please check again the content of `/etc/sudoers.d/` and run the following commands to see everything is ok with that file?
```
# stat /etc/sudoers.d/cis_sudoers
  File: /etc/sudoers.d/cis_sudoers
  Size: 17        	Blocks: 8          IO Block: 4096   regular file
Device: 805h/2053d	Inode: 790384      Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-06-25 08:23:21.070016366 +0200
Modify: 2021-06-25 08:22:51.838016477 +0200
Change: 2021-06-25 08:22:51.838016477 +0200
 Birth: -
# cat /etc/sudoers.d/cis_sudoers
Defaults use_pty
```

Regards,
Chema.

**Comment by fraform-tnz** [comment]
Hi @chemamartinez and thanks for reply.

Here's the **stat** and the content of cis_sudoers file (recreated just for you :P), but the check still fails. Just for clarify I tried also with other permissions mask but there is no change in check.

```
# stat cis_sudoers
  File: cis_sudoers
  Size: 17              Blocks: 8          IO Block: 4096   regular file
Device: 10303h/66307d   Inode: 1294        Links: 1
Access: (0440/-rw-r-----)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-06-28 07:09:35.317477125 +0000
Modify: 2021-06-28 07:09:35.317477125 +0000
Change: 2021-06-28 07:09:35.333477253 +0000
```
```
# cat cis_sudoers
Defaults use_pty
```

```
2021/06/28 07:18:19 sca[19634] wm_sca.c:1923 at wm_sca_check_dir(): DEBUG: Considering directory entry '/etc/sudoers.d/cis_sudoers'
2021/06/28 07:18:19 sca[19634] wm_sca.c:1944 at wm_sca_check_dir(): DEBUG: Skipping directory entry '/etc/sudoers.d/cis_sudoers'
2021/06/28 07:18:19 sca[19634] wm_sca.c:1960 at wm_sca_check_dir(): DEBUG: Check result for dir '/etc/sudoers.d': 0
2021/06/28 07:18:19 sca[19634] wm_sca.c:1118 at wm_sca_do_scan(): DEBUG: Check directory rule result: 0
2021/06/28 07:18:19 sca[19634] wm_sca.c:1157 at wm_sca_do_scan(): DEBUG: Result for rule 'd:/etc/sudoers.d -> r:\.* -> r:^\s*\t*Defaults\s*\t*use_pty': 0
2021/06/28 07:18:19 sca[19634] wm_sca.c:1180 at wm_sca_do_scan(): DEBUG: Result for check id: 2526 'Ensure sudo commands use pty' -> 0
```

**Comment by chemamartinez** [comment]
Hi again @fraform-tnz,

I have discovered the cause of the problem. I was not able to reproduce it because I was testing the 4.2 version (which is in the pre-release stage) and it is solved in there.

When testing 4.1.5 I got the error you report.

This is caused due to a failure when processing the regex `\.*` as the check includes, so it doesn't match the file name and it is skipped. You can read more about this issue at https://github.com/wazuh/wazuh/issues/3583.

In addition, this pull request fixes the issue https://github.com/wazuh/wazuh/pull/7495 and it is included in 4.2.0 which will be released soon. I am sorry for the inconvenience.

On the other hand, you can make the check pass as a workaround by adding the line `Defaults use_pty` to the file `/etc/sudoers` directly.

Thank you for reporting this to us!

Best regards,
Chema.

**Comment by fraform-tnz** [comment]
Hi @chemamartinez,
I am glad that the issue is (auto)solved. So if you confirm that in 4.2 it works without problem I can close this issue.

Thanks again.
